Mathlib/Lean/Expr/Basic.lean

```diff
@@ -189,6 +189,23 @@ def getAppApps (e : Expr) : Array Expr :=
   let nargs := e.getAppNumArgs
   getAppAppsAux e (mkArray nargs dummy) (nargs-1)
 
+/-- Erase proofs in an expression by replacing them with `sorry`s.
+
+This function replaces all proofs in the expression
+and in the types that appear in the expression
+by `sorryAx`s.
+The resulting expression has the same type as the old one.
+
+It is useful, e.g., to verify if the proof-irrelevant part of a definition depends on a variable.
+-/
+def eraseProofs (e : Expr) : MetaM Expr :=
+  Meta.transform e
+    (pre := fun e => do
+      if (← Meta.isProof e) then
+        return .continue (← mkSyntheticSorry (← inferType e))
```

Comment:
This looks good to me.

One thing to be aware of is that `Meta.transform` doesn't instantiate metavariables. That's fine — you just need to be sure to `instantiateMVars` before running `eraseProofs` (assuming you're working with expressions that might have metavariables). There's a possible gotcha, where if you're using this function in a local context whose local variables might have metavariables, `inferType` here could result in metavariables.

I'm tempted to say "it's OK that this is sensitive to metavariables", and that this latter issue is unlikely to cause applications of this function issues. It only pops up if `inferType` results in a metavariable whose type contains a proof. I'm just writing this here to record the possibility.

Should we merge then?

bors d+

Reply:
Let's merge.
bors merge